### PR TITLE
 [WIP] ConditionalTransferableComponentsAdapter Implementation

### PR DESCRIPTION
**Pull request by YanisLalou**
Fixes: #96 
Paper: https://proceedings.mlr.press/v48/gong16.pdf
Mostly eq 7-8-9-10

### CODE REVIEW

File: skada/_subspace.py [hunk 1]
```diff
@@ -1021,3 +1025,341 @@ def TransferSubspaceLearning(
...
+
+                # grassmann_constraint = {
+                #     "type": "eq",
+                #     "fun": grassmann_constraint_function,
+                # }
+
+                (_, W, _, _), _ = torch_minimize(
+                    func_torch, (alpha, W, G, H), tol=self.tol, max_iter=1
+                )
+                W = torch.tensor(W, dtype=torch.float64)
```
Comment (YanisLalou) [inline skada/_subspace.py:1241]: Here any idea how to enforce the constraint to the torch minimiser ?

Comment (ambroiseodt) [reply skada/_subspace.py:1241]: hello, here are a couple of ideas:
- Use the dual of the optimization problem resulting in adding the constraint to the loss in Pytorch:
min_x  f(x) st g(x) = 0 <==> min_{x, lambda} f(x) + lambda * g(x)
here the constraint would be the norm of W^TW - Id

- Another idea could be to generate the matrices directly on the Grassmanian manifold:
 a) geomtorch: https://geotorch.readthedocs.io/en/latest/orthogonal/grassmannian.html 

b) torch: https://pytorch.org/docs/stable/generated/torch.nn.utils.parametrizations.orthogonal.html
Une autre methode, c'est de parametriser tes fonctions tq ce soit orthogonal avec 

- Maybe not directly related, but here is an interesting paper with code to optimize orthogonal matrices: 
https://proceedings.mlr.press/v151/ablin22a/ablin22a.pdf (code: https://github.com/pierreablin/landing)

**Comment by YanisLalou** [comment]
By guaranteeing that W is on the Grassmann manifold we loose performance in the `Subspace method example on subspace shift dataset` as illustrated here:

With constraint:
image

Without constraint:
image

Is it expected or is there an issue ?
